File: lima_linguisticdata/scripts/xmlforms.pl
#!/usr/bin/perl -s

# Copyright 2002-2013 CEA LIST
# SPDX-FileCopyrightText: 2022 CEA LIST <gael.de-chalendar@cea.fr>
#
# SPDX-License-Identifier: MIT

use utf8
binmode STDOUT, ":utf8";

open (SOURCE,"<:utf8","$ARGV[0]") || die ("Impossible d'ouvrir $ARGV[0]");
open (OUT,">>:utf8","$ARGV[1]") || die ("Impossible d'ouvrir $ARGV[1]");

if ($main::h || $main::help || scalar(@ARGV)!=2) {
  print "USAGE : xmlforms [OPTIONS] inputfile outputfile\n";
  print "where [OPTIONS] are :\n";
  print "  -h or -help : print usage\n";
  print "  -desacc=[yes|no] : specify desacc attribute for entries. default is none, that equals 'yes'\n";
  print "  -entryop=[add|replace|delete] : specify op attribute for entries. default is none, that equals 'add'\n";
  print "  -lingop=[add|replace|delete] : specify op attribute for linginfos. default is none, that equals 'add'\n";
}

my $form = "";
my $lemma = "";
my $norm = "";

my $count = 0;
my $icount = 0;

# Does not bufferize outputs on stdout, for seeing progress
my $oldfh = select(STDOUT);
$| = 1;
select($oldfh);


while (<SOURCE>)
{
  chop();
  s/&/&amp;/g;
  s/"/&quot;/g;
  s/</&lt;/g;
  s/>/&gt;/g;
  if ($_ eq "") { next;}
  @data=split(/	/);
  s/#.*//;  # to remove end of line comments
#   if (/#/)  { next;}  # pour autoriser les commentaires 
  if (/^\s*$/)  { next;}	# to ignore empty lines

 if (scalar(@data) !=4) {
    print "xmlform: Invalid line '$_'\n";
    next;
  }


  if ($data[0] ne $form) {
    $form=$data[0];
    if ($count > 0) {
      print OUT "  </i>\n";
      print OUT "</entry>\n";
    }
    print OUT "<entry k=\"$form\"";
    if ($main::desacc) {
      print OUT " desacc=\"$main::desacc\"";
    }
    if ($main::entryop) {
      print OUT " op=\"$main::entryop\"";
    }
    print OUT ">\n";
    $icount = 0;
    $count++;
    if (($count % 10000) == 0) {
      print "\r$count";
    }
  }
  if (($icount==0) || ($data[1] ne $lemma) || ($data[2] ne $norm)) {
    $lemma=$data[1];
    $norm=$data[2];
    if ($icount > 0) {
      print OUT "  </i>\n";
    }
    print OUT "  <i";
    if ($lemma ne "") {
      print OUT " l=\"$lemma\"";
    }
    if ($norm ne "") {
      print OUT " n=\"$norm\"";
    }
    if ($main::lingop) {
      print OUT " op=\"$main::lingop\"";
    }
    print OUT ">\n";
    $icount++;
  }
  print OUT "    <p v=\"$data[3]\"/>\n";
}
if ($count > 0) {
  print OUT "  </i>\n";
  print OUT "</entry>\n";
}
print "\n";

File: lima_linguisticdata/scripts/xmlforms.py
#!/usr/bin/env python3

# Copyright 2002-2024 CEA LIST
# SPDX-FileCopyrightText: 2024 CEA LIST <gael.de-chalendar@cea.fr>
#
# SPDX-License-Identifier: MIT

import argparse
import codecs

from tqdm import tqdm


def print_usage():
    print("USAGE : xmlforms [OPTIONS] inputfile outputfile")
    print("where [OPTIONS] are :")
    print("  -h or -help : print usage")
    print(
        "  -desacc=[yes|no] : specify desacc attribute for entries. "
        "default is none, that equals 'yes'"
    )
    print(
        "  -entryop=[add|replace|delete] : specify op attribute for entries. "
        "default is none, that equals 'add'"
    )
    print(
        "  -lingop=[add|replace|delete] : specify op attribute for linginfos. "
        "default is none, that equals 'add'"
    )


def main():
    parser = argparse.ArgumentParser(
        description="Process input and output files with options."
    )
    parser.add_argument("inputfile", type=str, help="Input file")
    parser.add_argument("outputfile", type=str, help="Output file")
    parser.add_argument(
        "-desacc",
        type=str,
        choices=["yes", "no"],
        help="Specify desacc attribute for entries. "
        'Default is none, that equals "yes"',
    )
    parser.add_argument(
        "-entryop",
        type=str,
        choices=["add", "replace", "delete"],
        help="Specify op attribute for entries. " 'Default is none, that equals "add"',
    )
    parser.add_argument(
        "-lingop",
        type=str,
        choices=["add", "replace", "delete"],
        help="Specify op attribute for linginfos. "
        'Default is none, that equals "add"',
    )

    args = parser.parse_args()

    with codecs.open(args.inputfile, "r", "utf-8") as source:
        with codecs.open(args.outputfile, "a", "utf-8") as out:
            form = ""
            lemma = ""
            norm = ""

            count = 0
            icount = 0

            lines = source.readlines()
            for line in tqdm(lines, desc="Processing lines", unit="line"):
                line = line.strip()
                line = (
                    line.replace("&", "&amp;")
                    .replace('"', "&quot;")
                    .replace("<", "&lt;")
                    .replace(">", "&gt;")
                )
                if line == "":
                    continue

                line = line.split("#")[0].strip()
                data = line.split("\t")
                if line == "" or len(data) != 4:
                    print(f"xmlform: Invalid line '{line}'")
                    continue

                if data[0] != form:
                    form = data[0]
                    if count > 0:
                        out.write("  </i>\n</entry>\n")
                    out.write(f'<entry k="{form}"')
                    if args.desacc:
                        out.write(f' desacc="{args.desacc}"')
                    if args.entryop:
                        out.write(f' op="{args.entryop}"')
                    out.write(">\n")
                    icount = 0
                    count += 1

                if icount == 0 or data[1] != lemma or data[2] != norm:
                    lemma = data[1]
                    norm = data[2]
                    if icount > 0:
                        out.write("  </i>\n")
                    out.write("  <i")
                    if lemma:
                        out.write(f' l="{lemma}"')
                    if norm:
                        out.write(f' n="{norm}"')
                    if args.lingop:
                        out.write(f' op="{args.lingop}"')
                    out.write(">\n")
                    icount += 1

                out.write(f'    <p v="{data[3]}"/>\n')

        if count > 0:
            out.write("  </i>\n</entry>\n")


if __name__ == "__main__":
    main()
